Task 1 — Data Loading & Exploration


In [7]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import seaborn as sns

In [8]:
#In this block we are importing the dataset and printing the top 10 rows
df = pd.read_csv("WA_Fn-UseC_-HR-Employee-Attrition.csv")
print("Starting Ten Rows:\n")
df.head(10)

Starting Ten Rows:



,Age,Attrition,BusinessTravel,DailyRate,Department,DistanceFromHome,Education,EducationField,EmployeeCount,EmployeeNumber,...,RelationshipSatisfaction,StandardHours,StockOptionLevel,TotalWorkingYears,TrainingTimesLastYear,WorkLifeBalance,YearsAtCompany,YearsInCurrentRole,YearsSinceLastPromotion,YearsWithCurrManager
0,41,Yes,Travel_Rarely,1102,Sales,1,2,Life Sciences,1,1,...,1,80,0,8,0,1,6,4,0,5
1,49,No,Travel_Frequently,279,Research & Development,8,1,Life Sciences,1,2,...,4,80,1,10,3,3,10,7,1,7
2,37,Yes,Travel_Rarely,1373,Research & Development,2,2,Other,1,4,...,2,80,0,7,3,3,0,0,0,0
3,33,No,Travel_Frequently,1392,Research & Development,3,4,Life Sciences,1,5,...,3,80,0,8,3,3,8,7,3,0
4,27,No,Travel_Rarely,591,Research & Development,2,1,Medical,1,7,...,4,80,1,6,3,3,2,2,2,2
5,32,No,Travel_Frequently,1005,Research & Development,2,2,Life Sciences,1,8,...,3,80,0,8,2,2,7,7,3,6
6,59,No,Travel_Rarely,1324,Research & Development,3,3,Medical,1,10,...,1,80,3,12,3,2,1,0,0,0
7,30,No,Travel_Rarely,1358,Research & Development,24,1,Life Sciences,1,11,...,2,80,1,1,2,3,1,0,0,0
8,38,No,Travel_Frequently,216,Research & Development,23,3,Life Sciences,1,12,...,2,80,0,10,2,3,9,7,1,8
9,36,No,Travel_Rarely,1299,Research & Development,27,3,Medical,1,13,...,2,80,2,17,3,2,7,7,7,7


In [9]:
rows , cols = df.shape
print(f"Number of Rows: {rows} \nNumber of Cols: {cols}")

Number of Rows: 1470 
Number of Cols: 35


In [10]:
#Identifying the target columns
target_col = 'Attrition'
print(f"Target Column: {target_col}")
count = df['Attrition'].value_counts()
print(f"\nAttrition Count:")
print(count)

Target Column: Attrition

Attrition Count:
Attrition
No     1233
Yes     237
Name: count, dtype: int64


In [11]:
#Printing The rate of attrition 
attrition_rate = (count/len(df))*100
print(attrition_rate)

Attrition
No     83.877551
Yes    16.122449
Name: count, dtype: float64


In [12]:
#Identifying the different types of columns
numerical_col = df.select_dtypes(include=['int64', 'float64']).columns
categorical_col = df.select_dtypes(include=['object']).columns
print(f'Number of Numerical Columns: {len(numerical_col)}')
print(f'Number of Categorical Columns: {len(categorical_col)}')

print("\nList of columns:")
print("\nNumerical Cols: ",list(numerical_col))
print("\nCategorical Cols: ",list(categorical_col))

Number of Numerical Columns: 26
Number of Categorical Columns: 9

List of columns:

Numerical Cols:  ['Age', 'DailyRate', 'DistanceFromHome', 'Education', 'EmployeeCount', 'EmployeeNumber', 'EnvironmentSatisfaction', 'HourlyRate', 'JobInvolvement', 'JobLevel', 'JobSatisfaction', 'MonthlyIncome', 'MonthlyRate', 'NumCompaniesWorked', 'PercentSalaryHike', 'PerformanceRating', 'RelationshipSatisfaction', 'StandardHours', 'StockOptionLevel', 'TotalWorkingYears', 'TrainingTimesLastYear', 'WorkLifeBalance', 'YearsAtCompany', 'YearsInCurrentRole', 'YearsSinceLastPromotion', 'YearsWithCurrManager']

Categorical Cols:  ['Attrition', 'BusinessTravel', 'Department', 'EducationField', 'Gender', 'JobRole', 'MaritalStatus', 'Over18', 'OverTime']


In [13]:
#Visiualizing the columns percentage
columns = [len(numerical_col),len(categorical_col)]
names = ["Numerical Cols", "Categorical Cols"]
plt.pie(columns,labels=names,autopct='%.1f%%')
plt.savefig('columns.png', dpi=300, bbox_inches='tight')
plt.show()


C:\Users\Hemangi\AppData\Local\Temp\ipykernel_16680\1198803259.py:6: UserWarning: FigureCanvasAgg is non-interactive, and thus cannot be shown
  plt.show()


We can observed in the pie chart above that the distribution is imbalanced as only 25.7% of the total employees said yes and the majority said no to attrition .

Task 2 — Data Cleaning & Preprocessing


In [16]:
#Identifying missing values
missing_values = df.isnull().sum()
print(missing_values)

# in the output we can see that there are no missing values and hence there is no need for handling 

Age                         0
Attrition                   0
BusinessTravel              0
DailyRate                   0
Department                  0
DistanceFromHome            0
Education                   0
EducationField              0
EmployeeCount               0
EmployeeNumber              0
EnvironmentSatisfaction     0
Gender                      0
HourlyRate                  0
JobInvolvement              0
JobLevel                    0
JobRole                     0
JobSatisfaction             0
MaritalStatus               0
MonthlyIncome               0
MonthlyRate                 0
NumCompaniesWorked          0
Over18                      0
OverTime                    0
PercentSalaryHike           0
PerformanceRating           0
RelationshipSatisfaction    0
StandardHours               0
StockOptionLevel            0
TotalWorkingYears           0
TrainingTimesLastYear       0
WorkLifeBalance             0
YearsAtCompany              0
YearsInCurrentRole          0
YearsSince

In [17]:
#Dropping columns which do not add any value to target
df.drop(
    columns=["EmployeeNumber", "Over18", "StandardHours"],
    inplace=True
)


In [18]:
#Converting attrition col to numerical 
df['Attrition'] = df['Attrition'].map({
    "No":0,
    "Yes":1
})

In [19]:
#One-Hot Encoding
X = df.drop("Attrition", axis=1)
y = df["Attrition"]

categorical_cols = X.select_dtypes(include="object").columns
X = pd.get_dummies(
    X,
    columns = categorical_cols,
    drop_first=True
)

In [20]:
from sklearn.preprocessing import StandardScaler

numeric_cols = X.select_dtypes(include=["int64","float64"]).columns
scaler = StandardScaler()

X[numeric_cols]= scaler.fit_transform(X[numeric_cols])

Task 3 — Exploratory Data Analysis — EDA



In [22]:
#Department vs Attrition
department_attrition = pd.crosstab(
    df["Department"],
    df["Attrition"],
    normalize="index"
) * 100

department_attrition

Attrition,0,1
Department,,
Human Resources,80.952381,19.047619
Research & Development,86.160250,13.839750
Sales,79.372197,20.627803


The Sales department has the highest attrition rate (20.63%), followed by Human Resources (19.05%). Research & Development has the lowest attrition rate (13.84%). This suggests that HR should prioritize retention efforts in the Sales department.

In [24]:
#JobRole vs Attrition
job_role_attrition = pd.crosstab(
    df["JobRole"],
    df["Attrition"],
    normalize="index"
) * 100

job_role_attrition

Attrition,0,1
JobRole,,
Healthcare Representative,93.129771,6.870229
Human Resources,76.923077,23.076923
Laboratory Technician,76.061776,23.938224
Manager,95.098039,4.901961
Manufacturing Director,93.103448,6.896552
Research Director,97.500000,2.500000
Research Scientist,83.904110,16.095890
Sales Executive,82.515337,17.484663
Sales Representative,60.240964,39.759036


The **Sales Representative** Role has the highest attrition rate (39.7%) followed by **Laboratory Technician** with the rate of (23.9%).Sales Representatives and Laboratory Technicians should be prioritized for retention efforts, as they exhibit significantly higher attrition rates than other job roles.

In [26]:
#MonthlyIncome vs Attrition
df.groupby("Attrition")["MonthlyIncome"].describe()

,count,mean,std,min,25%,50%,75%,max
Attrition,,,,,,,,
0,1233.0,6832.739659,4818.208001,1051.0,3211.0,5204.0,8834.0,19999.0
1,237.0,4787.092827,3640.210367,1009.0,2373.0,3202.0,5916.0,19859.0


Employees who left the company had a significantly lower monthly income than employees who stayed. The average monthly income of employees who left was 4,787.09, compared to 6,832.74 for employees who stayed. Similarly, the median monthly income was 3,202 for employees who left versus 5,204 for employees who stayed. This suggests that lower salary is associated with a higher likelihood of employee attrition.

In [28]:
#WorkLifeBalance vs Attrition
worklife_attrition = pd.crosstab(
    df["WorkLifeBalance"],
    df["Attrition"],
    normalize="index"
) * 100

worklife_attrition

Attrition,0,1
WorkLifeBalance,,
1,68.750000,31.250000
2,83.139535,16.860465
3,85.778275,14.221725
4,82.352941,17.647059


Employees with a Work-Life Balance rating of 1 had the highest attrition rate at 31.25%, which is nearly double that of employees with ratings 2 (16.86%) and 3 (14.22%). Employees with a rating of 3 had the lowest attrition rate, indicating that better work-life balance is associated with improved employee retention.

In [30]:
#YearsAtCompany vs Attrition
df.groupby("Attrition")["YearsAtCompany"].describe()

,count,mean,std,min,25%,50%,75%,max
Attrition,,,,,,,,
0,1233.0,7.369019,6.096298,0.0,3.0,6.0,10.0,37.0
1,237.0,5.130802,5.949984,0.0,1.0,3.0,7.0,40.0


Employees who left the company had an average tenure of 5.13 years, whereas employees who stayed had an average tenure of 7.37 years. The median tenure was 3 years for employees who left compared to 6 years for those who stayed. This indicates that employees are more likely to leave during the early years of their employment.


Task 4: Model Building

In [33]:
from sklearn.model_selection import train_test_split

X_train,X_test,y_train,y_test = train_test_split(
    X,y,test_size=0.2,random_state=42,stratify=y
)
#startify is used to presevere the class dist in both training & test data sets

In [34]:
#Training using Logistic Regression
from sklearn.linear_model import LogisticRegression

lr = LogisticRegression(
    class_weight="balanced",
    random_state =42,
    max_iter=1000
)

lr.fit(X_train,y_train)

LogisticRegression(class_weight='balanced', max_iter=1000, random_state=42)

In [35]:
#Training using Random Forest
from sklearn.ensemble import RandomForestClassifier

rf = RandomForestClassifier(
    class_weight="balanced",
    random_state=42
)

rf.fit(X_train,y_train)

RandomForestClassifier(class_weight='balanced', random_state=42)

In [36]:
#Training using Gradient Boosting
from sklearn.ensemble import GradientBoostingClassifier

gb = GradientBoostingClassifier(
    random_state=42
)

gb.fit(X_train,y_train)

GradientBoostingClassifier(random_state=42)

Task 5 — Model Evaluation


In [38]:
from sklearn.metrics import(
    classification_report,
    confusion_matrix,
    ConfusionMatrixDisplay,
    roc_auc_score,
    roc_curve
)


In [39]:
# we will make a function so that we dont have to calculate for each model
def model_eval(model,X_test,y_test,model_name):
    y_pred = model.predict(X_test)
    
    y_prob = model.predict_proba(X_test)[:,1]

    # precision = classification_report(y_test,y_pred,output_dict=True)["1"]["precision"]
    # recall = classification_report(y_test,y_pred,output_dict=True)["1"]["recall"]
    # f1 = classification_report(y_test,y_pred,output_dict=True)["1"]["f1-score"]

    result = classification_report(y_test,y_pred,output_dict=True)
    precision= result["1"]["precision"]
    recall= result["1"]["recall"]
    f1= result["1"]["f1-score"]
    roc_auc = roc_auc_score(y_test,y_prob)


    print(model_name)

    print("\nClassification Report")
    print(classification_report(y_test, y_pred))

    print("\nROC-AUC Score:", round(roc_auc,3))

    # Confusion Matrix
    cm = confusion_matrix(y_test, y_pred)

    disp = ConfusionMatrixDisplay(confusion_matrix=cm)
    disp.plot(cmap="Blues")

    plt.title(model_name)
    plt.show()

    return {
        "Model": model_name,
        "Precision": precision,
        "Recall": recall,
        "F1 Score": f1,
        "ROC-AUC": roc_auc
    }

In [40]:
#Results for Logistic Reg
lr_results = model_eval(
    lr,
    X_test,
    y_test,
    "Logistic Regression"
)

Logistic Regression

Classification Report
              precision    recall  f1-score   support

           0       0.92      0.77      0.84       247
           1       0.36      0.66      0.46        47

    accuracy                           0.76       294
   macro avg       0.64      0.72      0.65       294
weighted avg       0.83      0.76      0.78       294


ROC-AUC Score: 0.804


C:\Users\Hemangi\AppData\Local\Temp\ipykernel_16680\518523116.py:32: UserWarning: FigureCanvasAgg is non-interactive, and thus cannot be shown
  plt.show()


In [41]:
#Results for Random Forest

rf_results = model_eval(
    rf,
    X_test,
    y_test,
    "Random Forest"
)

Random Forest

Classification Report
              precision    recall  f1-score   support

           0       0.85      0.99      0.91       247
           1       0.50      0.06      0.11        47

    accuracy                           0.84       294
   macro avg       0.67      0.53      0.51       294
weighted avg       0.79      0.84      0.78       294


ROC-AUC Score: 0.775


C:\Users\Hemangi\AppData\Local\Temp\ipykernel_16680\518523116.py:32: UserWarning: FigureCanvasAgg is non-interactive, and thus cannot be shown
  plt.show()


In [42]:
#Results for Gradient Boosting

gb_results = model_eval(
    gb,
    X_test,
    y_test,
    "Gradient Boosting"
)

C:\Users\Hemangi\AppData\Local\Temp\ipykernel_16680\518523116.py:32: UserWarning: FigureCanvasAgg is non-interactive, and thus cannot be shown
  plt.show()


Gradient Boosting

Classification Report
              precision    recall  f1-score   support

           0       0.87      0.97      0.92       247
           1       0.59      0.21      0.31        47

    accuracy                           0.85       294
   macro avg       0.73      0.59      0.61       294
weighted avg       0.82      0.85      0.82       294


ROC-AUC Score: 0.794


In [43]:
#Comparison Data Frame
comparison = pd.DataFrame([
    lr_results,
    rf_results,
    gb_results
])
print(comparison)

                 Model  Precision    Recall  F1 Score   ROC-AUC
0  Logistic Regression   0.356322  0.659574  0.462687  0.803601
1        Random Forest   0.500000  0.063830  0.113208  0.775218
2    Gradient Boosting   0.588235  0.212766  0.312500  0.793695


In [44]:
comparison.sort_values(
    by="ROC-AUC",
    ascending=False
)

,Model,Precision,Recall,F1 Score,ROC-AUC
0,Logistic Regression,0.356322,0.659574,0.462687,0.803601
2,Gradient Boosting,0.588235,0.212766,0.312500,0.793695
1,Random Forest,0.500000,0.063830,0.113208,0.775218


Logistic Regression performs the best among the three models. It has the highest ROC-AUC score of 0.80 , it also has the highesr F1-Score(0.462) and the highest recall(0.660). All these indicate that logistic regression is the best-performing model.

In [46]:
#Calculating the most important features
importance = pd.DataFrame({
    "Feature": X.columns,
    "Importance": abs(lr.coef_[0])
})

importance = importance.sort_values(
    by="Importance",
    ascending=False
)
top10 = importance.head(10)
print(top10)

                             Feature  Importance
44                      OverTime_Yes    1.626169
24  BusinessTravel_Travel_Frequently    1.596964
35     JobRole_Laboratory Technician    1.571637
41      JobRole_Sales Representative    1.264294
38         JobRole_Research Director    1.112054
31              EducationField_Other    1.020279
25      BusinessTravel_Travel_Rarely    0.903722
43              MaritalStatus_Single    0.864502
34           JobRole_Human Resources    0.670991
17                 TotalWorkingYears    0.612480


Task 6 — Visualization

In [48]:
#Attrition Rates by dept 

plt.figure(figsize=(8,5))

sns.countplot(
    data = df,
    x="Department",
    hue="Attrition",
    palette=["#89CFF0", "#F3C3C2"]
)
plt.title("Attrition Rates by Department")
plt.xlabel("\n\nDepartment")
plt.ylabel("Number of Employees")
plt.savefig('att-by-dept.png', dpi=300, bbox_inches='tight')
plt.show()


C:\Users\Hemangi\AppData\Local\Temp\ipykernel_16680\3825454710.py:15: UserWarning: FigureCanvasAgg is non-interactive, and thus cannot be shown
  plt.show()


In [49]:
#Attrition by Job Role

sns.countplot(
    data = df,
    x="JobRole",
    hue="Attrition",
    palette=["#4F8FC0","#FFE3B3"]
)
plt.title("Attrition by Job Role")
plt.xlabel("\nJob Role")
plt.ylabel("Number of Employees")
plt.xticks(rotation=90)
plt.savefig('att-by-job.png', dpi=300, bbox_inches='tight')
plt.show()

C:\Users\Hemangi\AppData\Local\Temp\ipykernel_16680\91167702.py:14: UserWarning: FigureCanvasAgg is non-interactive, and thus cannot be shown
  plt.show()


In [50]:
#Monthly Income vs Attrition
sns.boxplot(
    data=df,
    x="Attrition",
    y="MonthlyIncome"
)
plt.title("Monthly Income vs Attrition")
plt.xlabel("Attrition")
plt.ylabel("Monthly Income")
plt.savefig('att-by-income.png', dpi=300, bbox_inches='tight')
plt.show()

C:\Users\Hemangi\AppData\Local\Temp\ipykernel_16680\2844423463.py:11: UserWarning: FigureCanvasAgg is non-interactive, and thus cannot be shown
  plt.show()


In [51]:
#Confusion matrix for the best performing model
lr_pred = lr.predict(X_test)
cm = confusion_matrix(y_test, lr_pred)

disp = ConfusionMatrixDisplay(confusion_matrix=cm, display_labels=["Stayed", "Label"])
disp.plot(cmap="Purples")

plt.title("Logistic Regression- Confusion Matrix")
plt.savefig('CM-LR.png', dpi=300, bbox_inches='tight')
plt.show()

C:\Users\Hemangi\AppData\Local\Temp\ipykernel_16680\1826539028.py:10: UserWarning: FigureCanvasAgg is non-interactive, and thus cannot be shown
  plt.show()


In [52]:
#Plotting the most important features
# top10 = importance.head(10)
plt.barh(
    top10["Feature"],
    top10["Importance"]
)
plt.show()

C:\Users\Hemangi\AppData\Local\Temp\ipykernel_16680\1348079730.py:7: UserWarning: FigureCanvasAgg is non-interactive, and thus cannot be shown
  plt.show()


In [53]:
#ROC CURVE PROBABILITY
lr_prob_pred = lr.predict_proba(X_test)[:,1]
rf_prob_pred = rf.predict_proba(X_test)[:,1]
gb_prob_pred = gb.predict_proba(X_test)[:,1]

In [54]:
lr_fpr, lr_tpr, _ = roc_curve(y_test, lr_prob_pred)
rf_fpr, rf_tpr, _ = roc_curve(y_test, rf_prob_pred)
gb_fpr, gb_tpr, _ = roc_curve(y_test, gb_prob_pred)

plt.figure(figsize=(8,6))

plt.plot(
    lr_fpr,
    lr_tpr,
    label=f"Logistic Regression (AUC={roc_auc_score(y_test, lr_prob_pred):.3f})",
    color="yellow"
)

plt.plot(
    rf_fpr,
    rf_tpr,
    label=f"Random Forest (AUC={roc_auc_score(y_test, rf_prob_pred):.3f})",
    color="lightblue"
)

plt.plot(
    gb_fpr,
    gb_tpr,
    label=f"Gradient Boosting (AUC={roc_auc_score(y_test, gb_prob_pred):.3f})",
    color="lightgreen"
)

plt.plot(
    [0,1],
    [0,1],
    linestyle="--",
    color="pink"

)

plt.xlabel("False Positive Rate")
plt.ylabel("True Positive Rate")
plt.title("ROC Curve Comparison")

plt.legend()
plt.savefig('roc-curve.png', dpi=300, bbox_inches='tight')

plt.show()

C:\Users\Hemangi\AppData\Local\Temp\ipykernel_16680\3173671574.py:43: UserWarning: FigureCanvasAgg is non-interactive, and thus cannot be shown
  plt.show()


Task 7 – HR Insights & Business Recommendations

According to thlogistic r Regression, the best predictors of attrition include **OverTime_Yes**, **BusinessTravel_Travel_Frequently** and **JobRole_Laboratory Technician**. Thus, it can be concluded that attrition is more likely to happen among employees who work overtime, frequently travel for their jobs, and work as Laboratory Technicians. Salary does not affect attrition because there are other factors connected with the work process (overtime, travels, and others) that affect the behavior of an employee. According to the EDA, the company should concentrate on the department/job role that has the biggest percentage of attritionA). To reduce the number of people who leave the company, the company needs to follow up with employees who work overtime and take some steps to relieve their load and improve their work-life balance. Moreover, the company needs to conduct retention and career development meetings with employees who are at risk. The limitation of this model lies in its reliance on historical data and inability to take into account personal factors, culture, motivation, and actions of employees in the future.

In [56]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score

# Load Dataset
data = pd.read_csv("WA_Fn-UseC_-HR-Employee-Attrition.csv")

# Convert Attrition into numeric values
data["Attrition"] = data["Attrition"].map({"Yes":1,"No":0})

# Encode categorical columns
data = pd.get_dummies(data, drop_first=True)

# Features and Target
X = data.drop("Attrition", axis=1)
y = data["Attrition"]

# Train-Test Split
X_train, X_test, y_train, y_test = train_test_split(
    X, y,
    test_size=0.2,
    random_state=42
)

# Model Training
model = RandomForestClassifier(random_state=42)
model.fit(X_train, y_train)

# Prediction
predictions = model.predict(X_test)

# Accuracy
accuracy = accuracy_score(y_test, predictions)

print("="*60)
print("EMPLOYEE ATTRITION PREDICTION SYSTEM")
print("="*60)
print(f"Total Employees       : {len(data)}")
print(f"Training Records      : {len(X_train)}")
print(f"Testing Records       : {len(X_test)}")
print(f"Prediction Accuracy   : {accuracy*100:.2f}%")
print("="*60)
print("SYSTEM STATUS : ATTRITION PREDICTION COMPLETED")
print("="*60)

EMPLOYEE ATTRITION PREDICTION SYSTEM
Total Employees       : 1470
Training Records      : 1176
Testing Records       : 294
Prediction Accuracy   : 87.76%
SYSTEM STATUS : ATTRITION PREDICTION COMPLETED


In [98]:
import pandas as pd
import joblib

from sklearn.model_selection import train_test_split
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import OneHotEncoder, StandardScaler

from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.ensemble import GradientBoostingClassifier

from sklearn.metrics import accuracy_score

# Load Dataset
df = pd.read_csv("WA_Fn-UseC_-HR-Employee-Attrition.csv")

# Encode Target Variable
df["Attrition"] = df["Attrition"].map({"Yes":1,"No":0})

X = df.drop("Attrition", axis=1)
y = df["Attrition"]

# Numerical & Categorical Columns
num_cols = X.select_dtypes(include=["int64"]).columns
cat_cols = X.select_dtypes(include=["object"]).columns

# Preprocessing
preprocessor = ColumnTransformer([
    ("num", StandardScaler(), num_cols),
    ("cat", OneHotEncoder(handle_unknown="ignore"), cat_cols)
])

# Train-Test Split
X_train, X_test, y_train, y_test = train_test_split(
    X, y,
    test_size=0.20,
    random_state=42
)

# Logistic Regression
lr_model = Pipeline([
    ("preprocessor", preprocessor),
    ("classifier", LogisticRegression(max_iter=1000))
])

lr_model.fit(X_train, y_train)

# Random Forest
rf_model = Pipeline([
    ("preprocessor", preprocessor),
    ("classifier", RandomForestClassifier(random_state=42))
])

rf_model.fit(X_train, y_train)

# Gradient Boosting
gb_model = Pipeline([
    ("preprocessor", preprocessor),
    ("classifier", GradientBoostingClassifier(random_state=42))
])

gb_model.fit(X_train, y_train)

# Prediction
prediction = rf_model.predict(X_test)

accuracy = accuracy_score(y_test, prediction)
# accuracy = accuracy_score(y_test, prediction)

model_info = {
    "model": rf_model,
    "accuracy": accuracy
}

joblib.dump(model_info, "trained_model.pkl")

print("Random Forest Accuracy :", accuracy)

# Save Best Model
joblib.dump(rf_model, "trained_model.pkl")

print("Phase II Successfully Implemented")

Random Forest Accuracy : 0.8775510204081632
Phase II Successfully Implemented


In [100]:
# -----------------------------
# System Validation
# -----------------------------

import pandas as pd
import joblib

# Load dataset
df = pd.read_csv("WA_Fn-UseC_-HR-Employee-Attrition.csv")

# Load trained model
model = joblib.load("trained_model.pkl")

# Check dataset
print("Dataset Shape:", df.shape)

# Check missing values
print(df.isnull().sum())

# Verify prediction
sample = df.drop("Attrition", axis=1).iloc[[0]]

prediction = model.predict(sample)

probability = model.predict_proba(sample)

print("Prediction:", prediction)

print("Probability:", probability)

Dataset Shape: (1470, 35)
Age                         0
Attrition                   0
BusinessTravel              0
DailyRate                   0
Department                  0
DistanceFromHome            0
Education                   0
EducationField              0
EmployeeCount               0
EmployeeNumber              0
EnvironmentSatisfaction     0
Gender                      0
HourlyRate                  0
JobInvolvement              0
JobLevel                    0
JobRole                     0
JobSatisfaction             0
MaritalStatus               0
MonthlyIncome               0
MonthlyRate                 0
NumCompaniesWorked          0
Over18                      0
OverTime                    0
PercentSalaryHike           0
PerformanceRating           0
RelationshipSatisfaction    0
StandardHours               0
StockOptionLevel            0
TotalWorkingYears           0
TrainingTimesLastYear       0
WorkLifeBalance             0
YearsAtCompany              0
YearsInCurrent